### ISSUE TITLE
Support or document issues with ADLS (Azure Data Lake Storage) Gen 2

### ISSUE THREAD

**Issue by MigQ2** [treeverse/lakeFS#4931]
### TL;DR
**Using ADLS Gen2 (a.k.a Hierarchical Namespace) as underlying Storage for LakeFS is not fully supported and gives problems. Please document this in the Azure installation documentation**

## Full context and Details

I installed LakeFS using an Azure Storage Account with Hierarchical Namespace enabled and got an error when write multiple medium-sized files using Spark (please note I could read and write small files without error): `AmazonS3Exception: We encountered an internal error, please try again` (trying again didn't work, the error was consistent). 

Then in the LakeFS logs I found the following error: `FeatureNotYetSupportedForHierarchicalNamespaceAccounts FeatureName: Blob Tags`

I installed LakeFS in a different Azure Storage Account with Hierarchical Namespace disabled and everything started working fine.

## Extras

- Error happened when I tried to write to LakeFS from Azure Databricks using `s3a:/`
- LakeFS version: 0.86.0
- `blockstore.azure.auth_method` is "access-key"

**Comment by N-o-Z** [comment]
@MigQ2 Thanks for your contribution - please note that the following error is not a lakeFS error but comes from the Azure blob storage.
You are using a feature which is not supported in HNS mode.
Please take a look at the currently supported blob operations in HNS:
https://learn.microsoft.com/en-us/azure/storage/blobs/storage-feature-support-in-storage-accounts

**Comment by MigQ2** [comment]
Hi @N-o-Z, thanks for reaching out 

I am not using the Blog Tag feature explicitly myself, I guess at some point LakeFS is using it internally and that's how the error triggers. All I am doing is writing files to LakeFS `s3a` endpoint but I am not interacting directly with Azure Storage at all.

Hope this helps, let me know if I can share any additional info

**Comment by N-o-Z** [comment]
@MigQ2 Thanks - can you please provide the file sizes you are writing.
I will try to reproduce it and get back to you

**Comment by MigQ2** [comment]
It's not that large, around ~100MB compressed I would say. It has approx 2 million records and 100 columns. 

The code I was running when I got the error is as simple as: 

```
df = spark.read.load("abfss://mycontainer@myacct.dfs.core.windows.net/mypath", format="delta")
df.write.save("s3a://my-lakefs-repo/my-branch/somepath/", format="parquet")
```

**Comment by MigQ2** [comment]
@N-o-Z  I found this in the internal LakeFS logs in case it can help finding out where is LakeFS using Blob tags: 

```
time="2023-01-03T02:12:34Z" level=error msg="could not complete multipart upload" func="pkg/gateway/operations.(*PostObject).HandleCompleteMultipartUpload" file="build/pkg/gateway/operations/postobject.go:114"
[...]
PUT https://mystorageaccount.blob.core.windows.net/mycontainer/my-lakefs-repo/my-path/gqdnmkneua46cqfni2n0/cepou47eua46cqfni37g_id?comp=blocklist&timeout=601\n
```

It seems when lakeFS is doing the `postobject`, the PUT request to the Azure Blob API has a `comp=blocklist` parameter which [is the way of adding Blob Tags](https://learn.microsoft.com/en-us/rest/api/storageservices/set-blob-tags#request), concretely [blocklist](https://learn.microsoft.com/en-us/rest/api/storageservices/put-block-list#request). Any idea where can this `blocklist` be coming from?

**Comment by N-o-Z** [comment]
@MigQ2 Thanks for the info - it was actually very helpful.
The error is due to the multipart upload mechanism that uses tags for the list of parts in order to complete the upload.
This behavior is used by the azure blob sdk we are currently using which is probably incompatible with ADLS Gen2.
We will examine the possibility of using the new azure sdk - and see if this enables us to support HNS.
In any case we will add documentation stating the requirements for Azure and other block storage providers (You can follow up in https://github.com/treeverse/lakeFS/issues/4932)
Will try to give an update early next week

**Comment by N-o-Z** [comment]
@MigQ2 handled in:
- https://github.com/treeverse/lakeFS/pull/5111
- https://github.com/treeverse/lakeFS/issues/5037

Closing this issue
